src/Elasticsearch/Index/FieldDefinition.php

```diff
@@ -0,0 +1,271 @@
+  }
+
+  /**
+   * Return a property.
+   *
+   * @param $field_name
+   *
+   * @return \Drupal\elasticsearch_helper\Elasticsearch\Index\FieldDefinition|null
+   */
+  public function getProperty($field_name) {
+    return isset($this->properties[$field_name]) ? $this->properties[$field_name] : NULL;
+  }
+
+  /**
+   * Returns object properties.
+   *
+   * @return \Drupal\elasticsearch_helper\Elasticsearch\Index\FieldDefinition[]
+   */
+  public function getProperties() {
+    return $this->properties;
+  }
+
+  /**
+   * Returns TRUE if field has properties.
+   *
+   * @return bool
+   */
+  public function hasProperties() {
+    return !empty($this->properties);
+  }
+
+  /**
+   * Adds a multi-field.
+   *
+   * @param $field_name
+   * @param \Drupal\elasticsearch_helper\Elasticsearch\Index\FieldDefinition $field
+   *
+   * @return self
+   */
+  public function addMultiField($field_name, FieldDefinition $field) {
+    $this->validateMultiFieldAddition();
+
+    $this->multiFields[$field_name] = $field;
+
+    return $this;
+  }
+
+  /**
+   * Return a multi-field.
+   *
+   * @param $field_name
+   *
+   * @return \Drupal\elasticsearch_helper\Elasticsearch\Index\FieldDefinition|null
+   */
+  public function getMultiField($field_name) {
+    return isset($this->multiFields[$field_name]) ? $this->multiFields[$field_name] : NULL;
+  }
+
+  /**
+   * Returns properties.
+   *
+   * @return \Drupal\elasticsearch_helper\Elasticsearch\Index\FieldDefinition[]
+   */
+  public function getMultiFields() {
+    return $this->multiFields;
+  }
+
+  /**
+   * Returns TRUE if field has multi-fields.
+   *
+   * @return bool
+   */
+  public function hasMultiFields() {
+    return !empty($this->multiFields);
+  }
+
+  /**
+   * Returns field definition as an array.
+   *
+   * @param \Drupal\elasticsearch_helper\Elasticsearch\Index\FieldDefinition|NULL $source
+   *
+   * @return array
+   */
+  public function toArray(FieldDefinition $source = NULL) {
+    $source = $source ?: $this;
+
+    $result = $source->getOptions();
+
+    // Add a type.
+    $result['type'] = $source->getDataType()->getType();
+
+    // Include properties.
+    if ($source->hasProperties()) {
+      foreach ($source->getProperties() as $field_name => $field_definition) {
+        $result['properties'][$field_name] = $this->toArray($field_definition);
+      }
+    }
+    // Include multi-fields into the array if available.
+    elseif ($source->hasMultiFields()) {
+      foreach ($source->getMultiFields() as $field_name => $field_definition) {
+        $result['fields'][$field_name] = $this->toArray($field_definition);
+      }
+    }
+
+    return $result;
+  }
+
+  /**
+   * {@inheritdoc}
+   *
+   * @throws \InvalidArgumentException
+   */
+  protected function validateOptions(array $options = []) {
+    if (isset($options['type'])) {
+      throw new \InvalidArgumentException(t('Data type should be added as an argument to the constructor of @class.', [
```

Comment:
According to latest changes to coding standards, exception messages should not be translated. See https://www.drupal.org/docs/develop/coding-standards/php-exceptions#conventions

Thus, remove t() from all exceptions.